Add build_all script to Builds

MarkusTeufelberger opened the pull request:
This script will compile rippled, run unit tests and then delete all build artifacts in many possible configurations (in Linux).

-j4 is added to make it compile a bit faster. All commands are chained via &&, so if any of them fail, the script will stop at the error.

There are a couple more configurations possible by combining switches by the way, however if even just these would compile + run unit tests without any issue/error/warning for every release, it would already be amazing. :-)

I know that Test.py also exists, that one seems to build much fewer variants though and does not clean up between builds. I am not too much into shell scripting, but I guess this could also be rewritten in a few for-loops, if you are worried that it might slowly deprecate over time (like Test.py).

MarkusTeufelberger commented:
Any comments, anyone actually interested in this?

sublimator commented:
Thanks @MarkusTeufelberger  

I'm sure it's appreciated, just that people have their noses down so to speak.

sublimator commented:
@nbougalis @ximinez

sublimator commented:
There's a custom of tagging two reviewers for pull requests via `@`. I pinged two rippled team members for you :)

MarkusTeufelberger commented:
Yeah, since your team roles are not public, it's kinda hard to guess for me who is responsible for builds. ;-)

sublimator commented:
Yeah, just trying to explain/excuse the (as yet!) lack of response :)

ximinez commented:
There's a lot of overlap here with `Builds/Test.py`.

seelabs commented:
@MarkusTeufelberger Thought I'd share a bash snippet. This will get the number of physical cores: `num_procs=$(lscpu -p | grep -v '^#' | sort -u -t, -k 2,4 | wc -l) # pysical cores`

That lets you do `-j${num_procs}` rather than hard coding `4`.

MarkusTeufelberger commented:
@ximinez - yes and no. "Test.py --all" just builds "debug" and "release" for every available compiler. Not coverage, not the nounity variants, not profile and it doesn't know about extra options (and their combinations) at all. It (as I wrote above) also does not run the equivalent of "make clean" which actually seems to be an issue with the (failing for me) ninja builds. It does however also run npm tests... though I think you are currently working on replacing them with pure C++ ones, so I didn't add that so far.

@seelabs - I tried something like that but ran out of memory on some builds (I used a simpler version so it detected 8 cores instead of the 4 physical ones due to hyperthreading), thus the artificial static limit. Thanks for the suggestion, I'll add it in since it reports only the 4 physical cores on my machine.

ximinez commented:
@MarkusTeufelberger I know `Test.py` doesn't do exactly this. What I'm suggesting is that instead of creating a new script that isn't cross platform, and only works if you have both compilers installed, you modify `Test.py` to add the features you need.

Examples:
* `--universe` - Build `debug release coverage profile`
* `--clean` - Remove the build folder in between each target. Somehow.
* `--ninja` - Run `ninja` after `scons`.
* `--nonpm` - Skip running the npm tests.

You don't need to implement anything `--static`, `--assert`, or `--sanitize`, since those can already be passed through to `scons`.

MarkusTeufelberger commented:
I also thought about extending Test.py and making this script just call Test.py to build all possible combinations or even extending Test.py to a point where "--all" really builds everything. Would you prefer this approach?

The issue I have from time to time is that rippled breaks on some builds but not others and I'm not 100% sure if it is an issue on my end (e.g. the recent switch to gcc 6.1 on my machine messed up Boost pre-1.61 and https://llvm.org/bugs/show_bug.cgi?id=23529 is a constant source of pain...) or if it is something that needs to be fixed upstream. I thought if I gave you an easy way to run _all_ builds + tests, you might be more motivated to run them before tagging releases at least (running this on Travis would be overkill I guess).

ximinez commented:
@MarkusTeufelberger Yeah, I would prefer something that works with what we already have. ie. `Test.py`.

If you're getting crashes and/or segfaults when running the unit tests, it would be great if you could send us more information about them. I suspect the problem is more likely to be on our end than yours, since you've successfully been building rippled for so long. Building more configs on Travis is probably a non-starter, since it takes so long already.

MarkusTeufelberger commented:
Lately it's been mostly warnings or complete build failures on my end (e.g. rippled won't build at all until the new boost version is available, since it is the only one that builds with gcc 6.1). If I ran into segfaults or similar, I'd of course report them.

Alright, I'll work over Test.py a bit, I like Python more than bash anyways... :-)

MarkusTeufelberger commented:
Alright, I changed it a bit to reduce potential shadowing conflicts and apparently Test.py already does build unity and nounity by default. I added the power sets of possible options (afaik. you can't set both sanitizers at the same time?) to --all and a (preliminary) --clean switch.

You might(?) be able to shoot yourself in the foot if you try hard enough by defining build options yourself when invoking it with --all enabled, other than that it should at least be ready for some sweet compiling action.

Oh well, back to my actual homework. :-/

nbougalis commented:
Thanks @MarkusTeufelberger - I like `--clean` a lot (I find myself doing `rm -rf build/` a lot and I may or may not have accidentally removed `Builds` instead.

ximinez commented:
@MarkusTeufelberger This is definitely more on the right track! Unfortunately, the new build-all it isn't working very well on Windows - some of the build types aren't supported, and Windows scons seems to be doing something weird with the extra command line options. I've started fixing the issues as I've found them, but won't be able to put much time into this today. In the meantime, here's my working branch, so you can pick it up if you have access to a Windows build box, otherwise I'll get back to it later today or tomorrow. https://github.com/ximinez/rippled/commits/mt-build-all

ximinez commented:
Oh, one other suggestion. Why not support `--clean` in the regular builds, too?

MarkusTeufelberger commented:
I tried building rippled for Windows a few years ago, found out that the performant database backends (back then) were Linux only and installing VS is close to installing a whole different operating system on top of Windows... so I never really went back there (also because I couldn't find reasonably fast disk pooling options for Windows, so the storage backend is another issue). Is rippled on Windows really a thing out in the wild? The only package (after several years of rippled development) is RPM and I doubt that many people would bother to compile it on their own for Windows.

Build types not being supported (e.g. --sanitize=address on Windows) would probably just make scons itself fail with an error message I guess/hope. I'm not so sure if there is some proper way of getting a list of actually supported build features out of scons besides just trying or hardcoding them in some way (Test.py has no idea about compiler detection for starters). Maybe it would be easier to just use the -k switch when testing all possible builds?

To support --clean in regular builds, I'd need to rework Test.py a bit more, which I'll do probably over the weekend. Currently I tried to keep the existing code path intact and just add my own stuff in a different function to ensure it acts the same, in case some of your internal work flows depend on it. If a more or less complete rewrite/refactoring is ok with you, I'd also get rid of the globals for example.

I'll also add a --nonpm/-n switch to only run the C++ unit tests instead of both npm and C++, since I currently seem to run into issues with file descriptor counts and outdated/deprecated packages in the deeper dependencies of some npm modules.

vinniefalco commented:
I run rippled on Windows almost exclusively and it works great with NuDB on SSD.

nbougalis commented:
I'm fine with this. :+1:

ximinez commented:
@MarkusTeufelberger It's not perfect, but I've made some changes to make it work under Windows (mainly by stripping stuff out), and resolving some issues under linux (Ubuntu). I still haven't had a complete successful run through, partly because it takes _*FOREVER*_. Have you?

MarkusTeufelberger commented:
For me it usually fails on some builds (I'm on Manjaro [basically Arch] Linux) by default because of Clang's/GCC's abi_tag issues, static libraries not being shipped while being hardcoded by scons, Boost going nuts with a new compiler or ninja builds failing alltogether because they don't seem to build the protobuf files for me.

The original (shell) script went through once for me overnight, but I had to disable several builds that were problematic on my platform, as noted above. I then wanted to give you guys the script first and afterwards maybe spin up a Ubuntu Docker instance to see if the issue is my distro (as you officially only support Ubuntu LTS) or the SConstruct file. I have quite a bit to do at the moment, so I unfortunately didn't get to it yet, but I'll still work on this pull request on the upcoming weekend. If you improved the script in the meantime (especially Windows, which I can't test), by all means please feel free to share your changes! :+1: 

Running it all will take a while for sure, such is the price for having lots of options. Still I think it is important to actually run everything that you want to support, even if it might be slight overkill or rather used as a "hindsight" tool (e.g. run it on every beta release, but don't delay the release until you've finished the run - but be prepared to push out a new release or deliver quick patches if the build breaks). I would expect that at least RC versions had been built and unit-tested in every possible configuration. Because it is probably cumbersome to do so manually, I think this script is a reasonable way to just fire up in the evening and check the next day.
Another use case is probably checking if/how rippled builds on various Linux distros. Maybe one day you want to provide packages not only for RHEL...? Here a possible Docker option might come into play that fires up vanilla containers, installs dependencies and then tries to build/test everything here.

ximinez commented:
Oh, shoot, I forgot to include the link to my commit with the Windows changes. https://github.com/ximinez/rippled/commit/e0e4896f0ebdb0bf5debc28472e73ad5774d1a83

MarkusTeufelberger commented:
About build times: I checked and on my machine with -j4 it takes about 4-5 minutes for a full build from scratch + unit tests + npm tests. With a total of 16 (compiler + target + [no]unity) * 24 (all combinations of options) * 5 (minutes) this means that a full run of all possible builds will take about 32 hours. About half of that time is unit tests, the other half is compiling, so it might even be possible to cut this time in half just by compiling the next binary while still testing the current one. Otoh I'm using a stock i5-2500k, a newer PC with more cores and RAM might be faster anyways. Also this can be distributed anyways if time is critical, build + test are quite closed units.

MarkusTeufelberger commented:
Only "issue" for now that I'm aware of: The dot in the windows regex is not escaped, but it just matches the actual dot in its place.

Other than that I'd say the Test.py file can now be considered complete, I haven't tested it very extensively, but so far it seems to do what it is supposed to do, also on Python3 (bytes needed to be decoded to become strings).

ximinez commented on Builds/Test.py:
```diff
@@ -48,18 +48,35 @@
+# list of tuples of all possible options
+ALL_OPTIONS = list(set(
+    [tuple(x) for x in powerset(['--ninja', '--static', '--assert', '--sanitize=address'])] +
+    [tuple(x) for x in powerset(['--ninja', '--static', '--assert', '--sanitize=thread'])]))
```
Only the `--assert` option is supported on Windows. I suggest
```
if IS_WINDOWS:
  ALL_OPTIONS = [tuple(x) for x in powerset(['--assert'])]
else:
  ...
```

MarkusTeufelberger replied:
Ah, sure!

ximinez commented on Builds/Test.py:
```diff
@@ -75,55 +95,69 @@
+        lines.append(line.decode())
         if verbose:
-            print(line, end='')
+            print(line.decode(), end='')
```
What is the need for these `decode()`s? It doesn't seem to be improving anything, and the :heavy_check_mark: characters in the npm output are making it puke in Windows builds.

MarkusTeufelberger replied:
Random first google result: http://pythoncentral.io/encoding-and-decoding-strings-in-python-3-x/

On my machine (using python3) all returned lines are actually byte arrays (https://www.python.org/dev/peps/pep-3112/), meaning the output is not formatted at all and just consists of b'foo\n',b'bar\n',b'baz\n' and so on. Windows traditionally has... issues with encodings. Which Python version, Windows version and Python distribution are you using? I'd like to keep the decode() in, if possible, maybe again with an "IS_WINDOWS" switch in front that sets an encoding that your machine uses.

ximinez replied:
The obvious difference is that I'm using Python 2.7.8 (from https://www.python.org/downloads/). Windows 7 SP1. Here's the error:
```
Traceback (most recent call last):
  File "./Builds/Test.py", line 265, in <module>
    main()
  File "./Builds/Test.py", line 247, in main
    failed = run_tests(args)
  File "./Builds/Test.py", line 202, in run_tests
    resultcode, lines = shell('npm', ('test', '--rippled=' + executable,))
  File "./Builds/Test.py", line 158, in shell
    lines.append(line.decode())
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 5: ordinal not in range(128)
```
And here's what fixed it for me:
```
index 071d641..c4d9541 100755
--- a/Builds/Test.py
+++ b/Builds/Test.py
@@ -155,9 +155,13 @@ def shell(cmd, args=(), silent=False):
     lines = []
     count = 0
     for line in process.stdout:
-        lines.append(line.decode())
+        if isinstance(line, str):
+            decoded = line
+        else:
+            decoded = line.decode()
+        lines.append(decoded)
         if verbose:
-            print(line.decode(), end='')
+            print(decoded, end='')
         elif not silent:
             count += 1
             if count >= 80:
```

MarkusTeufelberger replied:
Thanks, added.

ximinez commented:
I think we've finally got it. :+1:

nbougalis commented:
Merged as 55377c12d34bed58daf0c10ce66b334f9eeeea4a.
